ingestion/src/metadata/profiler/metrics/system/system.py

```diff
@@ -58,207 +38,81 @@ def recursive_dic():
-                    dataset_id=dataset_id,
-                    project_id=project_id,
-                    insert=DatabaseDMLOperations.INSERT.value,
-                    update=DatabaseDMLOperations.UPDATE.value,
-                    delete=DatabaseDMLOperations.DELETE.value,
-                    merge=DatabaseDMLOperations.MERGE.value,
-                )
-            )
-        )
-        jobs = [
-            BigQueryQueryResult(
-                query_type=row.statement_type,
-                timestamp=row.start_time,
-                table_name=row.destination_table,
-                dml_statistics=row.dml_statistics,
-            )
-            for row in cursor_jobs
-        ]
-        set_cache(
-            SYSTEM_QUERY_RESULT_CACHE,
-            f"{Dialects.BigQuery}.{project_id}.{dataset_id}.jobs",
-            jobs,
+T = TypeVar("T")
+
+
+class CacheProvider(ABC, Generic[T]):
+    def __init__(self):
+        self.cache = LRUCache[T](LRU_CACHE_SIZE)
+
+    def get_or_update_cache(
+        self,
+        cache_path: str,
+        get_queries_fn: Callable[..., List[T]],
+        *args,
+        **kwargs,
+    ):
+        if cache_path in self.cache:
+            return self.cache.get(cache_path)
+        result = get_queries_fn(*args, **kwargs)
+        self.cache.put(cache_path, result)
+        return result
+
+
+class EmptySystemMetricsSource:
+    """Empty system metrics source that can be used as a default. Just returns an empty list of system metrics
+    for any resource."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs.pop("session", None)
+        if len(args) > 0:
+            logger.warning("Received unexpected arguments: %s", args)
+        if len(kwargs) > 0:
+            logger.warning("Received unexpected keyword arguments: %s", kwargs)
+        super().__init__()
+
+    def get_inserts(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get insert queries"""
+        return []
+
+    def get_deletes(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get delete queries"""
+        return []
+
+    def get_updates(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get update queries"""
+        return []
+
+    def get_kwargs(self, *args, **kwargs):
+        return {}
+
+
+class SystemMetricsComputer(EmptySystemMetricsSource):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def get_system_metrics(self, *args, **kwargs) -> List[SystemProfile]:
+        """Return system metrics for a given table. Actual passed object can be a variety of types based
+        on the underlying infrastructure. For example, in the case of SQLalchemy, it can be a Table object
+        and in the case of Mongo, it can be a collection object."""
+        kwargs = super().get_kwargs(*args, **kwargs)
+        return (
+            super().get_inserts(**kwargs)
+            + super().get_deletes(**kwargs)
+            + super().get_updates(**kwargs)
```

Comment:
Why do we have this very loose interface and we are not enforcing specific interface that children should implement.

Reply:
This is a dependency injection pattern that uses `super`. This method which is type-safe and deterministic. More information in [this video](https://www.youtube.com/watch?v=EiOglTERPEo).

```
                                                        
                                                        
              ┌───────────────────────┐                 
              │                       │                 
              │  DependencyInterface  │                 
              │                       │                 
              └──────▲───────▲────────┘                 
                     │       │                          
                     │       │                          
    ┌──────────────┬─┘       └──┬────────────────────┐  
    │              │            │                    │  
    │   Procedure  │            │ ConcreteDependency │  
    │              │            │                    │  
    └───────────▲──┘            └──▲─────────────────┘  
                │                  │                    
                │                  │                    
               ┌┴─────────────────┬┘                    
               │                  │         ▼           
               │  Implementation  │                     
               │                  │                     
               └──────────────────┘                     
```

Reply:
```
from metadata.mixins.sqalchemy.sqa_mixin import SQAInterfaceMixin
class MyClass(SQAInterfaceMixin):
    pass
```
But this is a MixIn right, so it is never intended to be used like this -- a MixIn will never be used in single inheritance? 

To get back to the initial question, why can't we have 

```
def get_system_metrics(self):
    inserts = self.get_inserts(...)
    deletes = self.get_deletes(...)
    updates = self.get_updates(...)
    return SystemMetrics()

@abastrcatclass
def get_inserts(self):
    raise NoteImplementedError

...
```
With strong interface children need to implement

Reply:
My goal was to enable extending functionality of classes and dependency injection while keeping the system type safe, deterministic and modularized. More specifically:

1. We have a common interface but we dont want to couple how we get the data (example: sqlalchemy) with the framework's logic. 
2. We have some functionality is required at the bottom of the diamond (see diagram). Notice how they all require the sql_alchemy session? With linear inheritance (like you suggested) all the implementing children will need to get this sql alchemy session even though some will not need it (or need another object). We are currently doing this using the `runner` object and the `table` object which are not type safe (the profiler needs to guess what object its getting).
3. Note how each system metrics source implements the cache provider with a different generic type and uses it differently. This framework allows us to create an independent state in each object but have common functionality. I dont see how this can be achieved with a `CacheMixin` but I am open to suggestions.
4. The solution we currently have in place is the `SQAInterfaceMixin` which is (1) not type-safe, (2) uses `self` which disables functionality changes downstream, any change requires changing the parent class and therefore **all the children** and it has no method of managing state as it relies on the state of the inheriting child.
5. This is the idea of dependency injection. Each implementing class takes what it needs to accomplish the task and is independent of how others are accomplishing it.

Regarding your example. I fail to notice how it achieves the following:
1. Allowing default cases for classes that do not need to implement the system profile interface (all profilers need to inherit it).
2. Allowing common functionality (example: using the `Metrics` objects) while allowing different methods for retrieving the data.
3. Enabling state for implementing children while unifying functionality. Using `utils` will not achieve this because as the `utils` module grows one needs to guess with operations are relevant for the object one is developing. This method limits the available functionality to the scope of the what's available for the object.